File: javascript/aliceAndBobGame.js
//  https://www.hackerrank.com/challenges/alice-and-bobs-silly-game/problem
//  find how many prime numbers there are between 2 and n
function sillyGame(n) {
	if (n === 1) return 'Bob';

	isPrime = new Array(n + 1).fill(true);
	let num = 2;
	// mark non prime numbers
	while (num * num <= n) {
		if (isPrime[num]) {
			for (i = num * num; i < n + 1; i += num) {
				isPrime[i] = false;
			}
		}
		num += 1;
	}
	const count = isPrime.reduce((c, p) => (p ? c + 1 : c),0);

	return count%2 ? 'Alice' : 'Bob'
}

console.log(sillyGame(11));

File: python/aliceAndBobGame.py
# https://www.hackerrank.com/challenges/alice-and-bobs-silly-game/problem
# find how many prime numbers there are between 2 and n
def sillyGame(n):
    if n == 1:
        return 'Bob'
    
    is_prime = [1] * (n+1)
    num =2
    # mark non prime numbers
    while num*num <= n:
        if is_prime[num]:
            for i in range(num*num,n+1,num):
                is_prime[i] = 0
        num += 1

    return 'Alice' if is_prime[2:].count(1)%2 else 'Bob'
   